xmtp_api_d14n/src/middleware/aggregate_client.rs

```diff
@@ -0,0 +1,331 @@
+use crate::d14n::{GetNodes, HealthCheck};
+use futures::StreamExt;
+use prost::bytes::Bytes;
+use std::collections::HashMap;
+use std::num::NonZeroUsize;
+use thiserror::Error;
+use xmtp_api_grpc::{client::GrpcClient, error::GrpcError};
+use xmtp_common::{
+    RetryableError,
+    time::{Duration, Instant},
+};
+use xmtp_proto::{
+    ApiEndpoint,
+    api::{ApiClientError, BodyError, Client, Query},
+    prelude::ApiBuilder,
+};
+
+/* AggregateClient struct and impls */
+
+#[derive(Debug, Error)]
+pub enum AggregateClientError {
+    #[error("all node clients failed to build")]
+    AllNodeClientsFailedToBuild,
+    #[error(transparent)]
+    BodyError(#[from] BodyError),
+    #[error(transparent)]
+    GrpcError(#[from] ApiClientError<GrpcError>),
+    #[error("no nodes found")]
+    NoNodesFound,
+    #[error("no responsive nodes found under {latency}ms latency")]
+    NoResponsiveNodesFound { latency: u64 },
+    #[error("timeout reaching node {} under {}ms latency", node_id, latency)]
+    TimeoutNode { node_id: u32, latency: u64 },
+    #[error("unhealthy node {}", node_id)]
+    UnhealthyNode { node_id: u32 },
+}
+
+/// From<AggregateClientError> for ApiClientError<E> is used to convert the AggregateClientError to an ApiClientError.
+/// Required by the Client trait implementation, as request and stream can return AggregateClientError.
+impl<E> From<AggregateClientError> for ApiClientError<E>
+where
+    E: std::error::Error + Send + Sync + 'static,
+{
+    fn from(value: AggregateClientError) -> ApiClientError<E> {
+        ApiClientError::<E>::Other(Box::new(value))
+    }
+}
+
+/// RetryableError for AggregateClientError is used to determine if the error is retryable.
+/// Trait needed by the From<AggregateClientError> for ApiClientError<E> implementation.
+/// All errors are not retryable.
+impl RetryableError for AggregateClientError {
+    fn is_retryable(&self) -> bool {
+        false
```

Comment:
going to need to passthrough is_retryable to errors in the enum which might be retryable, like the rest of ApiClientError or BodyError, i.e

```rust
fn is_retryable(&self) -> bool {
    use AggregateClientError::*;
    match self {
        GrpcError(e) => e.is_retryable(),
        BodyError(e) => e.is_retryable(),
    }
}
```